File: prisma/refs/py/trapping-rain-water.py
def trap(height):
    if not height:
        return 0
    l, r = 0, len(height) - 1
    max_l = max_r = out = 0
    while l < r:
        if height[l] < height[r]:
            max_l = max(max_l, height[l])
            out += max_l - height[l]
            l += 1
        else:
            max_r = max(max_r, height[r])
            out += max_r - height[r]
            r -= 1
    return out

File: prisma/refs/trapping-rain-water.ts
export function trap(height: number[]): number {
  let l = 0, r = height.length - 1;
  let maxL = 0, maxR = 0, out = 0;
  while (l < r) {
    if (height[l] < height[r]) {
      maxL = Math.max(maxL, height[l]);
      out += maxL - height[l];
      l++;
    } else {
      maxR = Math.max(maxR, height[r]);
      out += maxR - height[r];
      r--;
    }
  }
  return out;
}
